theories/datatypes/List.ec

```diff
@@ -1818,6 +1819,113 @@ lemma mapiP x0 (f : int -> 'a -> 'b) (s : 'a list) y :
+op put (s : 'a list) (n : int) (x : 'a) =
+  if (0 <= n < size s)
+  then take n s ++ (x :: drop (n + 1) s)
+  else s.
+
+lemma put_empty (n : int) (x : 'a) :
+  put [] n x = [].
+proof. by smt(). qed.
+
+lemma size_put (s : 'a list) (n : int) (x : 'a) :
+  size (put s n x) = size s.
+proof.
+rewrite /put; case (0 <= n < size s) => [[ge0_n lts_n] | //].
+by rewrite size_cat size_take // lts_n /= size_drop /#.
+qed.
+
+lemma put_inrng (s : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s => put s n x = take n s ++ (x :: drop (n + 1) s).
+proof. by rewrite /put => ->. qed. 
+
+lemma put_outrng (s : 'a list) (n : int) (x : 'a) :
+  !(0 <= n < size s) => put s n x = s.
+proof. by rewrite /put => ->. qed. 
+
+lemma put_cons_0 (s : 'a list) (x y : 'a) :
+  put (y :: s) 0 x = x :: s.
+proof. by rewrite /put /= drop0 addrC ltzS size_ge0. qed. 
+  
+lemma put_cons (s : 'a list) (n : int) (x y : 'a) :
+  1 <= n => put (y :: s) n x = y :: (put s (n - 1) x).
+proof. move => ge1_n @/put /#. qed.
+
+lemma put_cat_inrngfirst (s1 s2 : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s1 => put (s1 ++ s2) n x = put s1 n x ++ s2.
+proof.
+move => [ge0_n lts1_n]; rewrite /put size_cat ge0_n lts1_n /=.
+have ->: n < size s1 + size s2.
++ case (size s2 = 0) => [-> //|]; rewrite -lt0n 1:size_ge0 => ge0_s2. 
+  by rewrite (ltz_trans (size s1)) 2:ltz_addl. 
+rewrite /= take_cat lts1_n /= drop_cat. 
+case (n + 1 < size s1) => [|/lezNgt len1_s1]. 
++ by rewrite -cat_cons catA //.
++ rewrite -lez_add1r addrC in lts1_n.
+  have eqn1s1: n + 1 = size s1 by rewrite (lez_anti (size s1) (n + 1)).
+  by rewrite eqn1s1 subzz drop0 drop_size -cat1s catA.
+qed.
+
+lemma put_cat_inrngsecond (s1 s2 : 'a list) (n : int) (x : 'a) :
+  size s1 <= n < size s1 + size s2 => put (s1 ++ s2) n x = s1 ++ put s2 (n - size s1) x.
+proof.
+move => [ges1_n lts1s2_n]; rewrite /put size_cat.
+have ge0_n: 0 <= n by rewrite (lez_trans (size s1)).
+have ge0_ns1: 0 <= n - size s1 by rewrite subz_ge0.
+have lts2_ns1: n - size s1 < size s2 by smt().
+rewrite ge0_n lts1s2_n ge0_ns1 lts2_ns1 /= take_cat drop_cat.
+by rewrite ltzNge ges1_n (_ : n + 1 < size s1 = false) 1:/# /= catA addrAC.
+qed.
+ 
+lemma nth_put_inrng (dflt : 'a) (s : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s => nth dflt (put s n x) n = x.
+proof.
+elim: s n => [/# | y s ih n rng_n]; case (n = 0) => [-> | neq0_n].
++ by rewrite put_cons_0.
++ by rewrite put_cons 1:/# /= neq0_n /= ih 1:/#. 
+qed.
+
+lemma nth_put_outrng (dflt : 'a) (s : 'a list) (n : int)  (x : 'a) :
+  !(0 <= n < size s) => nth dflt (put s n x) n = dflt.
+proof. by move=> rng_n; rewrite put_outrng 2:nth_out. qed.
+
+lemma mem_put (s : 'a list) (n : int)  (x : 'a) :
+  0 <= n < size s => mem (put s n x) x.
+proof.
+by move => rng_n; rewrite /put rng_n /= mem_cat; right; rewrite in_cons.
+qed.
+
+lemma map_put (f : 'a -> 'b) (s : 'a list) (n : int) (x : 'a) :
+  map f (put s n x) = put (map f s) n (f x).
+proof.
+case (0 <= n < size s) => [| rng_n].
```

Comment:
This is a direct consequence of the properties of `map` and the definition of `put`. The induction seems superfluous here.